templates/components/engines/python/chat/engine.py

```diff
@@ -1,22 +1,32 @@
 import os
 
-from app.engine.index import get_index
+from app.engine.index import IndexConfig, get_index
 from app.engine.node_postprocessors import NodeCitationProcessor
 from fastapi import HTTPException
+from llama_index.core.callbacks import CallbackManager
 from llama_index.core.chat_engine import CondensePlusContextChatEngine
+from llama_index.core.memory import ChatMemoryBuffer
+from llama_index.core.settings import Settings
 
 
-def get_chat_engine(filters=None, params=None):
+def get_chat_engine(filters=None, params=None, event_handlers=None):
     system_prompt = os.getenv("SYSTEM_PROMPT")
     citation_prompt = os.getenv("SYSTEM_CITATION_PROMPT", None)
     top_k = int(os.getenv("TOP_K", 0))
+    llm = Settings.llm
+    memory = ChatMemoryBuffer.from_defaults(
+        token_limit=llm.metadata.context_window - 256
+    )
+    callback_manager = CallbackManager(handlers=event_handlers or [])
+    llm.callback_manager = callback_manager
```

Comment:
isn't Settings global? 
1. why do we need to set the callback_manager for the llm?
2. what happens to llm's callback_manager in multi-threading?

Reply:
same for agent/engine.py